### PR TITLE
Initial commit for sdk

### PR DESCRIPTION
**Pull request by Hemangi-Chauhan**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,168 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>org.dft</groupId>
```
Comment (dftrakesh) [inline pom.xml]: it should be fixed io.github.dftrakesh

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    <artifactId>VeracrossSDK</artifactId>
```
Comment (dftrakesh) [inline pom.xml]: should be veracross-sdk

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        <version.plugin.maven-gpg-plugin>1.6</version.plugin.maven-gpg-plugin>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.13.4</version>
+        </dependency>
+        <dependency>
```
Comment (dftrakesh) [inline pom.xml:44]: Change dependency to

<dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
            <version>1.18.24</version>
        </dependency>

File: src/main/java/com/dft/veracross/VeracrossParents.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import com.dft.veracross.model.parents.ParentsWrapper;
+import com.dft.veracross.model.parents.parentWrapper.ParentWrapper;
+import lombok.SneakyThrows;
+
+import java.net.URI;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.HashMap;
+
+public class VeracrossParents extends VeracrossSDK {
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossParents.java]: Rename class name to Parents

File: src/main/java/com/dft/veracross/VeracrossParents.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    private static final String FORWARD_SLASH_CHARACTER = "/";
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossParents.java]: Move this constant code to VeracrossSDK

File: src/main/java/com/dft/veracross/VeracrossParents.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    public VeracrossParents(VeracrossCredentials credentials) {
+        super(credentials);
+    }
+
+    public ParentWrapper getParentById(Integer id, VeracrossCredentials veracrossCredentials) {
+        URI uri = baseUrl(PARENTS_ENDPOINT.concat(FORWARD_SLASH_CHARACTER)
+                .concat(id.toString()));
+
+        System.out.println("uri: " + uri);
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossParents.java]: Remove System out use @Log4j2

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+@AllArgsConstructor
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: Change it to ReuqiredArgsConstrctor

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+@Builder(builderMethodName = "newBuilder", toBuilder = true)
+public class VeracrossSDK {
+
+    private final VeracrossCredentials credentials;
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: Arrange all fileds with similar type

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    int MAX_ATTEMPTS = 50;
+    int TIME_OUT_DURATION = 60000;
+    private final HttpClient client;
+    private static final String AUTHORIZATION = "Authorization";
+    private static final String BASE_ENDPOINT = "https://api.veracross.com/api-sandbox/v3";
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: By default it should be production URL
Add one constructor to set like isSanbox. If its passed then URL will be changed to Sandbox.

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            }
+            queryString.append(keyValueParam);
+        }
+
+        String pathWithQueryString = uri.getPath();
+        if (queryString.length() > 0) {
+            pathWithQueryString += "?" + queryString.toString();
+        }
+
+        return new URI(uri.getScheme(), uri.getAuthority(), pathWithQueryString, uri.getFragment());
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: Use URI.create

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    public VeracrossSDK(VeracrossCredentials credentials) {
+        this.credentials = credentials;
+        client = HttpClient.newBuilder()
+                .followRedirects(HttpClient.Redirect.ALWAYS)
+                .version(HttpClient.Version.HTTP_1_1)
+                .connectTimeout(Duration.ofSeconds(20))
+                .build();
+    }
+
+    @SneakyThrows
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: Remove @SneakyThrows after changing new URI to URI.create 
check next comment.

Comment (dftrakesh) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: Do same for other methods.

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    protected URI baseUrl(String path) {
+        return new URI(new StringBuilder()
+                .append(BASE_ENDPOINT)
+                .append(path)
+                .toString());
+    }
+
+    @SneakyThrows
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: What error it is throwing ?

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                .body();
+    }
+
+    @SneakyThrows
+    public <T> CompletableFuture<HttpResponse<T>> tryResend(HttpClient client,
+                                                            HttpRequest request,
+                                                            HttpResponse.BodyHandler<T> handler,
+                                                            HttpResponse<T> resp, int count) {
+        if (resp.statusCode() == 429 && count < MAX_ATTEMPTS) {
+            Thread.sleep(TIME_OUT_DURATION);
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: https://api-docs.veracross.com/docs/docs/ZG9jOjEyNzQyNjkz-rate-limiting

Implement rate limiting as per documenttaion. Check shipstation SDK for reference.

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+            return client.sendAsync(request, handler)
+                    .thenComposeAsync(response -> tryResend(client, request, handler, response, count + 1));
+        }
+        return CompletableFuture.completedFuture(resp);
+    }
+
+    public VeracrossParents getParentApi() {
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: Remove this method

File: src/main/java/com/dft/veracross/VeracrossSDK.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        return new VeracrossParents(credentials);
+    }
+
+    public VeracrossStudents getStudentApi() {
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/VeracrossSDK.java]: Remove this method

**Review by dftrakesh — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 1 above)
Comment (Hemangi-Chauhan) [reply pom.xml]: Change done.

File: pom.xml (hunk 2 above)
Comment (Hemangi-Chauhan) [reply pom.xml]: Changed to veracross-sdk.

File: src/main/java/com/dft/veracross/VeracrossParents.java (hunk 4 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossParents.java]: > Rename class name to Parents

Name changed to Parent.

File: src/main/java/com/dft/veracross/VeracrossParents.java (hunk 5 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossParents.java]: > Move this constant code to VeracrossSDK

Moved.

File: src/main/java/com/dft/veracross/VeracrossParents.java (hunk 6 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossParents.java]: > Remove System out use @log4j2

Removed unnecessary code from everywhere.

File: src/main/java/com/dft/veracross/VeracrossSDK.java (hunk 8 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: > Arrange all fileds with similar type

Arranged with similar field.

File: src/main/java/com/dft/veracross/VeracrossSDK.java (hunk 9 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: > By default it should be production URL Add one constructor to set like isSanbox. If its passed then URL will be changed to Sandbox.

Now i added sandbox in creadential class, now scool rout pass through credential. so not get issue when it comes to production url.

File: src/main/java/com/dft/veracross/VeracrossSDK.java (hunk 10 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: > 
Changed to URI.create

File: src/main/java/com/dft/veracross/VeracrossSDK.java (hunk 13 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: > https://api-docs.veracross.com/docs/docs/ZG9jOjEyNzQyNjkz-rate-limiting
> 
> Implement rate limiting as per documenttaion. Check shipstation SDK for reference.

Implemented as per documentation

File: src/main/java/com/dft/veracross/VeracrossSDK.java (hunk 14 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: > Remove this method

Removed.

File: src/main/java/com/dft/veracross/VeracrossSDK.java (hunk 15 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: > Remove this method

Removed.

File: pom.xml (hunk 3 above)
Comment (dftrakesh) [reply pom.xml:44]: Remove  <scope>compile</scope>

File: src/main/java/com/dft/veracross/credentials/VeracrossCredentials.java [hunk 16]
```diff
@@ -0,0 +1,16 @@
+package com.dft.veracross.credentials;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/credentials/VeracrossCredentials.java]: Why do you need @NoArgsConstructor

File: src/main/java/com/dft/veracross/model/parents/Parents.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    public String firstName;
+    public String firstNickName;
+    public Object preferredName;
+    public Object middleName;
+    public String lastName;
+    public Integer nameSuffix;
+    public Spouse spouse;
+    public Integer gender;
+    public String roles;
+    public Object email1;
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/model/parents/Parents.java]: It should String not object

File: src/main/java/com/dft/veracross/model/parents/parentWrapper/Address.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.Data;
+
+@Data
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+public class Address {
+
+    public String address1;
+    public Object address2;
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/model/parents/parentWrapper/Address.java]: Object ?

File: src/main/java/com/dft/veracross/model/parents/parentWrapper/Address.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+    public Object address3;
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/model/parents/parentWrapper/Address.java]: Object ?

File: test
Comment (dftrakesh) [inline test:1]: Remove TEST file.

File: src/main/java/com/dft/veracross/model/students/studentWrapper/StudentInfo.java [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import lombok.Data;
+
+@Data
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/model/students/studentWrapper/StudentInfo.java]: Do you need this annotation for this class ?

File: src/main/java/com/dft/veracross/model/students/studentWrapper/StudentInfo.java [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import lombok.Data;
+
+@Data
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/model/students/studentWrapper/StudentInfo.java:9]: Same here

File: src/main/java/com/dft/veracross/model/students/studentWrapper/StudentData.java [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+@Data
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+public class StudentData {
+
+    public Integer id;
+    public Integer namePrefix;
+    public String firstName;
+    public Object middleName;
```
Comment (dftrakesh) [inline src/main/java/com/dft/veracross/model/students/studentWrapper/StudentData.java]: Provide proper data type please. for all Object

**Review by dftrakesh — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 3 above)
Comment (Hemangi-Chauhan) [reply pom.xml:44]: > Remove compile

Removed

File: src/main/java/com/dft/veracross/VeracrossSDK.java (hunk 11 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/VeracrossSDK.java]: Changed to URI.create and removed @SneakyThrows.

File: src/main/java/com/dft/veracross/model/parents/Parents.java (hunk 17 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/model/parents/Parents.java]: > It should String not object

Set proper datatype everywhere as per data response

File: src/main/java/com/dft/veracross/model/students/studentWrapper/StudentData.java (hunk 22 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/model/students/studentWrapper/StudentData.java]: > Provide proper data type please. for all Object

Provided proper datatype in everyclass.

File: src/main/java/com/dft/veracross/model/students/studentWrapper/StudentInfo.java (hunk 20 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/model/students/studentWrapper/StudentInfo.java]: > Do you need this annotation for this class ?

No, i removed.

File: src/main/java/com/dft/veracross/credentials/VeracrossCredentials.java (hunk 16 above)
Comment (Hemangi-Chauhan) [reply src/main/java/com/dft/veracross/credentials/VeracrossCredentials.java]: > Why do you need @NoArgsConstructor

Sir, i used for testing in test class.

**Pull request merged**